metrics/prometheus.go

```diff
@@ -0,0 +1,263 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/raystack/raccoon/config"
+	"github.com/raystack/raccoon/logger"
+	"github.com/spf13/cast"
+)
+
+type CounterVec interface {
+	With(labels prometheus.Labels) prometheus.Counter
+	Collect(chan<- prometheus.Metric)
+	Describe(chan<- *prometheus.Desc)
+}
+
+type GaugeVec interface {
+	With(labels prometheus.Labels) prometheus.Gauge
+	Collect(chan<- prometheus.Metric)
+	Describe(chan<- *prometheus.Desc)
+}
+
+type HistogramVec interface {
+	With(labels prometheus.Labels) prometheus.Observer
+	Collect(chan<- prometheus.Metric)
+	Describe(chan<- *prometheus.Desc)
+}
+
+type PrometheusCollector struct {
+	registry  *prometheus.Registry
+	counters  map[string]CounterVec
+	gauges    map[string]GaugeVec
+	histogram map[string]HistogramVec
+	server    *http.Server
+}
+
+func initPrometheusCollector() (*PrometheusCollector, error) {
+	serveMux := &http.ServeMux{}
+	server := &http.Server{Addr: fmt.Sprintf(":%d", config.MetricPrometheus.Port), Handler: serveMux}
+	p := &PrometheusCollector{
+		counters:  getCounterMap(),
+		gauges:    getGaugeMap(),
+		histogram: getHistogramMap(),
+		registry:  prometheus.NewRegistry(),
+		server:    server,
+	}
+	err := p.Register()
+	if err != nil {
+		return nil, err
+	}
+	serveMux.Handle(config.MetricPrometheus.Path, promhttp.HandlerFor(p.registry, promhttp.HandlerOpts{Registry: p.registry}))
+	go server.ListenAndServe()
+	return p, nil
+
+}
+
+func (p *PrometheusCollector) Count(metricName string, count int64, labels map[string]string) error {
+	counter, ok := p.counters[metricName]
+	if !ok {
+		return fmt.Errorf("invalid counter metric %s", metricName)
```

Comment:
Dynamic setting of counters would be a code smell, since all metrics as well as label keys are immutable once registered. And any divergence results in a panic. So, if we adopt such a pattern where counters are dynamically registered. 
It could result in future metrics being used directly without initialisation.
If calls to such a metric is done with different or insufficient label keys. It would result in a panic.

Reply:
ok